File: leetcode/36/validSudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        tmp = [defaultdict(int) for _ in range(9)]
        spot = [defaultdict(int) for _ in range(9)] 
        for i in range(9):
            dic = defaultdict(int)
            for j in range(9):
                ri = i//3 
                rj = j//3
                r = ri+ri*2 + rj

                if board[i][j]!= ".":
                    dic[board[i][j]] += 1
                    spot[r][board[i][j]] += 1
                    if dic[board[i][j]] > 1:
                        # print(1, dic)
                        return False
                    if spot[r][board[i][j]] > 1:
                        # print(3, tmp)
                        return False
                if board[j][i] != ".":
                    tmp[i][board[j][i]] += 1
                    if tmp[i][board[j][i]] > 1:
                        # print(2, tmp)
                        return False
                
        # print(4, tmp)
        return True

        
        
        

File: leetcode/36/validSudoku.swift
class Solution {
    func isValidSudoku(_ board: [[Character]]) -> Bool {
        // 행, 열, 및 박스를 추적하는 Set을 생성
        var rows = Array(repeating: Set<Character>(), count: 9)
        var columns = Array(repeating: Set<Character>(), count: 9)
        var boxes = Array(repeating: Set<Character>(), count: 9)
        
        for i in 0..<9 {
            for j in 0..<9 {
                let char = board[i][j]
                if char == "." {
                    continue
                }
                // 행 중복 체크
                if rows[i].contains(char) {
                    return false
                }
                // 열 중복 체크
                if columns[j].contains(char) {
                    return false
                }
                // 박스 중복 체크
                let boxIndex = (i / 3) * 3 + j / 3
                if boxes[boxIndex].contains(char) {
                    return false
                }
                // 현재 문자 추가
                rows[i].insert(char)
                columns[j].insert(char)
                boxes[boxIndex].insert(char)
            }
        }
        
        return true
    }
}
